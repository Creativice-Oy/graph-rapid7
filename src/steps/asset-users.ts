import {
  createDirectRelationship,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { entities, relationships, steps } from '../constants';
import { getUserKey } from './access';

export async function fetchAssetUsers({
  logger,
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, logger);

  await jobState.iterateEntities(
    { _type: entities.ASSET._type },
    async (assetEntity) => {
      await apiClient.iterateAssetUsers(
        assetEntity.id as string,
        async (user) => {
          const userEntity = await jobState.findEntity(getUserKey(user.id));

          if (userEntity) {
            await jobState.addRelationship(
              createDirectRelationship({
                _class: RelationshipClass.OWNS,
                from: userEntity,
                to: assetEntity,
              }),
            );
          }
        },
      );
    },
  );
}

export const assetUsersStep: IntegrationStep<IntegrationConfig>[] = [
  {
    id: steps.FETCH_ASSET_USERS,
    name: 'Fetch Asset Users',
    entities: [],
    relationships: [relationships.USER_OWNS_ASSET],
    dependsOn: [steps.FETCH_ASSETS, steps.FETCH_USERS],
    executionHandler: fetchAssetUsers,
  },
];
