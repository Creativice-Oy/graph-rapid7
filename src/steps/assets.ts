import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../types';
import { entities, steps } from '../constants';

export function getAssetKey(id: number): string {
  return `insightvm_asset:${id}`;
}

export async function fetchAssets({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  await apiClient.iterateAssets(async (asset) => {
    const webLink = asset.links.find((link) => link.rel === 'self')?.href;

    const assetEntity = createIntegrationEntity({
      entityData: {
        source: asset,
        assign: {
          _key: getAssetKey(asset.id),
          _type: entities.ASSET._type,
          _class: entities.ASSET._class,
          id: `${asset.id}`,
          name: asset.hostName,
          category: 'server',
          make: 'unknown',
          model: 'unknown',
          serial: 'unknown',
          webLink,
        },
      },
    });

    await jobState.addEntity(assetEntity);
  });
}

export const assetsSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: steps.FETCH_ASSETS,
    name: 'Fetch Assets',
    entities: [entities.ASSET],
    relationships: [],
    executionHandler: fetchAssets,
  },
];
