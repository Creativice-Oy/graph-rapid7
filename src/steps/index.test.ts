import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { Recording, setupRapid7Recording } from '../../test/helpers/recording';

import { IntegrationConfig } from '../types';
import { fetchUsers } from './access';
import { fetchAccountDetails } from './account';
import { fetchSites } from './sites';
import { fetchScans } from './scans';
import { fetchAssets } from './assets';
import { fetchSiteAssets } from './site-assets';
import { fetchAssetVulnerabilities } from './vulnerabilities';
import { fetchAssetUsers } from './asset-users';
import { fetchScanAssets } from './scan-assets';

const DEFAULT_INSIGHT_HOST = 'localhost:3780';
const DEFAULT_INSIGHT_CLIENT_USERNAME = 'admin';
const DEFAULT_INSIGHT_CLIENT_PASSWORD = 'admin-password';

const integrationConfig: IntegrationConfig = {
  insightHost: process.env.INSIGHT_HOST || DEFAULT_INSIGHT_HOST,
  insightClientUsername:
    process.env.INSIGHT_CLIENT_USERNAME || DEFAULT_INSIGHT_CLIENT_USERNAME,
  insightClientPassword:
    process.env.INSIGHT_CLIENT_PASSWORD || DEFAULT_INSIGHT_CLIENT_PASSWORD,
};

jest.setTimeout(10 * 1000);

describe('Rapid7 InsightVM', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupRapid7Recording({
      directory: __dirname,
      name: 'insightvm_recordings',
      options: {
        recordFailedRequests: true,
      },
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    // Simulates dependency graph execution.
    // See https://github.com/JupiterOne/sdk/issues/262.
    await fetchAccountDetails(context);
    await fetchUsers(context);
    await fetchSites(context);
    await fetchScans(context);
    await fetchAssets(context);
    await fetchAssetUsers(context);
    await fetchSiteAssets(context);
    await fetchScanAssets(context);
    await fetchAssetVulnerabilities(context);

    // Review snapshot, failure is a regression
    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Account'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Account'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_account' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          webLink: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          accessUrl: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('User'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['User'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_user' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          username: {
            type: 'string',
          },
          email: {
            type: 'string',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Site'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Site'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_site' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          assets: {
            type: 'number',
          },
          importance: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          type: {
            type: 'string',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Process'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Process'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_scan' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          assets: {
            type: 'number',
          },
          siteName: {
            type: 'string',
          },
          engineName: {
            type: 'string',
          },
          name: {
            type: 'string',
          },
          state: {
            type: 'string',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Device'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Device'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_asset' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: {
            type: 'string',
          },
          category: {
            type: 'string',
          },
          make: {
            type: 'string',
          },
          model: {
            type: 'string',
          },
          serial: {
            type: 'string',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Finding'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Finding'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_finding' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: {
            type: 'string',
          },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter((e) =>
        e._class.includes('Vulnerability'),
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Vulnerability'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'insightvm_vulnerability' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: {
            type: 'string',
          },
          category: {
            type: 'string',
          },
          severity: {
            type: 'string',
          },
          blocking: {
            type: 'boolean',
          },
          open: {
            type: 'boolean',
          },
          production: {
            type: 'boolean',
          },
          public: {
            type: 'boolean',
          },
          webLink: {
            type: 'string',
          },
        },
      },
    });
  });
});
