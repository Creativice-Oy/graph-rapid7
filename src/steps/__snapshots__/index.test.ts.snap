// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rapid7 InsightVM should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "insightvm_account:root",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "cpu": Object {
              "clockSpeed": 4226,
              "count": 8,
            },
            "disk": Object {
              "free": Object {
                "bytes": 5900726272,
                "formatted": "5.5 GB",
              },
              "installation": Object {
                "backups": Object {
                  "bytes": 0,
                  "formatted": "0 bytes",
                },
                "database": Object {
                  "bytes": 2927711450,
                  "formatted": "2.7 GB",
                },
                "directory": "/opt/rapid7/nexpose",
                "reports": Object {
                  "bytes": 0,
                  "formatted": "0 bytes",
                },
                "scans": Object {
                  "bytes": 45477,
                  "formatted": "44.4 KB",
                },
                "total": Object {
                  "bytes": 5528381438,
                  "formatted": "5.1 GB",
                },
              },
              "total": Object {
                "bytes": 82268987392,
                "formatted": "76.6 GB",
              },
            },
            "distinguishedName": "CN=Rapid7 Security Console, O=CodeWorkr",
            "fqdn": "extreme-desktop",
            "host": "extreme-desktop",
            "ip": "127.0.1.1",
            "jvm": Object {
              "name": "OpenJDK 64-Bit Server VM",
              "startTime": "2020-10-05T11:26:08.492Z",
              "uptime": "PT11H50M36.584S",
              "vendor": "Azul Systems, Inc.",
              "version": "25.252-b14",
            },
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/administration/info",
                "rel": "self",
              },
            ],
            "memory": Object {
              "free": Object {
                "bytes": 167944192,
                "formatted": "160.2 MB",
              },
              "total": Object {
                "bytes": 16684572672,
                "formatted": "15.5 GB",
              },
            },
            "operatingSystem": "LinuxMint Linux 19.3",
            "serial": "A34F9A111E1CEC0D693E2B1E0DCB6E6AD737F5F0",
            "superuser": true,
            "user": "root",
            "version": Object {
              "build": "213",
              "changeset": "fa14553838f1251ad008d86c906520c8bbfff55c",
              "platform": "Linux64",
              "semantic": "6.6.46",
              "update": Object {
                "content": "1500556400",
                "contentPartial": "3503989404 (2020-10-02)",
                "id": Object {
                  "productId": "299067162755562",
                  "versionId": "490",
                },
                "product": "505267850",
              },
            },
          },
        },
      ],
      "_type": "insightvm_account",
      "accessUrl": "localhost:3780",
      "createdOn": undefined,
      "displayName": "root",
      "name": "root",
      "webLink": "https://localhost:3780/api/3/administration/info",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "insightvm_user:1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "authentication": Object {
              "external": false,
              "id": 1,
              "links": Array [
                Object {
                  "href": "https://localhost:3780/api/3/authentication_sources/1",
                  "rel": "self",
                },
                Object {
                  "href": "https://localhost:3780/api/3/authentication_sources/1/users",
                  "rel": "Authentication Source Users",
                },
              ],
              "name": "Builtin Administrators",
              "type": "admin",
            },
            "email": "contact@codeworkr.dev",
            "enabled": true,
            "id": 1,
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/users/1",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/authentication_sources/1",
                "rel": "Authentication Source",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/1/asset_groups",
                "rel": "Asset Groups",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/1/sites",
                "rel": "Sites",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/1/privileges",
                "rel": "Privileges",
              },
              Object {
                "href": "https://localhost:3780/api/3/roles/global-admin",
                "rel": "Role",
              },
            ],
            "locale": Object {
              "default": "en-US",
              "reports": "en-US",
            },
            "locked": false,
            "login": "admin",
            "name": "admin",
            "role": Object {
              "allAssetGroups": true,
              "allSites": true,
              "id": "global-admin",
              "name": "Global Administrator",
              "privileges": Array [
                "all-permissions",
                "approve-vulnerability-exceptions",
                "assign-scan-engine",
                "assign-scan-template",
                "assign-ticket-assignee",
                "close-tickets",
                "configure-global-settings",
                "create-reports",
                "create-tickets",
                "delete-vulnerability-exceptions",
                "manage-asset-group-access",
                "manage-asset-group-assets",
                "manage-dynamic-asset-groups",
                "manage-policies",
                "manage-report-access",
                "manage-report-templates",
                "manage-scan-alerts",
                "manage-scan-engines",
                "manage-scan-templates",
                "manage-site-access",
                "manage-site-credentials",
                "manage-sites",
                "manage-static-asset-groups",
                "manage-tags",
                "purge-site-asset-data",
                "schedule-automatic-scans",
                "specify-scan-targets",
                "specify-site-metadata",
                "start-unscheduled-scans",
                "submit-vulnerability-exceptions",
                "use-restricted-report-sections",
                "view-asset-group-asset-data",
                "view-site-asset-data",
              ],
              "superuser": true,
            },
          },
        },
      ],
      "_type": "insightvm_user",
      "createdOn": undefined,
      "displayName": "admin",
      "email": "contact@codeworkr.dev",
      "id": "1",
      "name": "admin",
      "username": "admin",
      "webLink": "https://localhost:3780/api/3/users/1",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "insightvm_user:2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "authentication": Object {
              "external": false,
              "id": 2,
              "links": Array [
                Object {
                  "href": "https://localhost:3780/api/3/authentication_sources/2",
                  "rel": "self",
                },
                Object {
                  "href": "https://localhost:3780/api/3/authentication_sources/2/users",
                  "rel": "Authentication Source Users",
                },
              ],
              "name": "Builtin Users",
              "type": "normal",
            },
            "email": "johndoe@codework.dev",
            "enabled": true,
            "id": 2,
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/users/2",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/authentication_sources/2",
                "rel": "Authentication Source",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/2/asset_groups",
                "rel": "Asset Groups",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/2/sites",
                "rel": "Sites",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/2/privileges",
                "rel": "Privileges",
              },
              Object {
                "href": "https://localhost:3780/api/3/roles/user",
                "rel": "Role",
              },
            ],
            "locale": Object {
              "default": "en-US",
              "reports": "en-US",
            },
            "locked": false,
            "login": "TestUser",
            "name": "John Doe",
            "role": Object {
              "allAssetGroups": false,
              "allSites": false,
              "id": "user",
              "name": "User",
              "privileges": Array [
                "assign-ticket-assignee",
                "close-tickets",
                "create-reports",
                "create-tickets",
                "view-asset-group-asset-data",
                "view-site-asset-data",
              ],
              "superuser": false,
            },
          },
        },
      ],
      "_type": "insightvm_user",
      "createdOn": undefined,
      "displayName": "John Doe",
      "email": "johndoe@codework.dev",
      "id": "2",
      "name": "John Doe",
      "username": "TestUser",
      "webLink": "https://localhost:3780/api/3/users/2",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "insightvm_user:3",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "authentication": Object {
              "external": false,
              "id": 1,
              "links": Array [
                Object {
                  "href": "https://localhost:3780/api/3/authentication_sources/1",
                  "rel": "self",
                },
                Object {
                  "href": "https://localhost:3780/api/3/authentication_sources/1/users",
                  "rel": "Authentication Source Users",
                },
              ],
              "name": "Builtin Administrators",
              "type": "admin",
            },
            "email": "read@creativice.com",
            "enabled": true,
            "id": 3,
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/users/3",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/authentication_sources/1",
                "rel": "Authentication Source",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/3/asset_groups",
                "rel": "Asset Groups",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/3/sites",
                "rel": "Sites",
              },
              Object {
                "href": "https://localhost:3780/api/3/users/3/privileges",
                "rel": "Privileges",
              },
              Object {
                "href": "https://localhost:3780/api/3/roles/global-admin",
                "rel": "Role",
              },
            ],
            "locale": Object {
              "default": "en-US",
              "reports": "en-US",
            },
            "locked": false,
            "login": "ReadOnlyUser",
            "name": "Read User",
            "role": Object {
              "allAssetGroups": true,
              "allSites": true,
              "id": "global-admin",
              "name": "Global Administrator",
              "privileges": Array [
                "all-permissions",
                "approve-vulnerability-exceptions",
                "assign-scan-engine",
                "assign-scan-template",
                "assign-ticket-assignee",
                "close-tickets",
                "configure-global-settings",
                "create-reports",
                "create-tickets",
                "delete-vulnerability-exceptions",
                "manage-asset-group-access",
                "manage-asset-group-assets",
                "manage-dynamic-asset-groups",
                "manage-policies",
                "manage-report-access",
                "manage-report-templates",
                "manage-scan-alerts",
                "manage-scan-engines",
                "manage-scan-templates",
                "manage-site-access",
                "manage-site-credentials",
                "manage-sites",
                "manage-static-asset-groups",
                "manage-tags",
                "purge-site-asset-data",
                "schedule-automatic-scans",
                "specify-scan-targets",
                "specify-site-metadata",
                "start-unscheduled-scans",
                "submit-vulnerability-exceptions",
                "use-restricted-report-sections",
                "view-asset-group-asset-data",
                "view-site-asset-data",
              ],
              "superuser": true,
            },
          },
        },
      ],
      "_type": "insightvm_user",
      "createdOn": undefined,
      "displayName": "Read User",
      "email": "read@creativice.com",
      "id": "3",
      "name": "Read User",
      "username": "ReadOnlyUser",
      "webLink": "https://localhost:3780/api/3/users/3",
    },
    Object {
      "_class": Array [
        "Site",
      ],
      "_key": "insightvm_site:1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "assets": 1,
            "id": 1,
            "importance": "normal",
            "lastScanTime": "2020-09-24T15:06:38.656Z",
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/sites/1",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/alerts",
                "rel": "Alerts",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/scan_engine",
                "rel": "Scan Engine",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/scan_schedules",
                "rel": "Schedules",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/organization",
                "rel": "Organization",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/tags",
                "rel": "Tags",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/users",
                "rel": "Users",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/scan_template",
                "rel": "Template",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/site_credentials",
                "rel": "Site Credentials",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/shared_credentials",
                "rel": "Assigned Shared Credentials",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/web_authentication/html_forms",
                "rel": "Web HTML Forms Authentication",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/web_authentication/http_headers",
                "rel": "Web HTTP Headers Authentication",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/assets",
                "rel": "Assets",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/included_targets",
                "rel": "Included Targets",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/excluded_targets",
                "rel": "Excluded Targets",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/included_asset_groups",
                "rel": "Included Asset Groups",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1/excluded_asset_groups",
                "rel": "Excluded Asset Groups",
              },
            ],
            "name": "Local Site",
            "riskScore": 578,
            "scanEngine": 3,
            "scanTemplate": "full-audit-without-web-spider",
            "type": "static",
            "vulnerabilities": Object {
              "critical": 0,
              "moderate": 1,
              "severe": 0,
              "total": 1,
            },
          },
        },
      ],
      "_type": "insightvm_site",
      "assets": 1,
      "createdOn": undefined,
      "displayName": "Local Site",
      "id": "1",
      "importance": "normal",
      "name": "Local Site",
      "type": "static",
      "webLink": "https://localhost:3780/api/3/sites/1",
    },
    Object {
      "_class": Array [
        "Assessment",
      ],
      "_key": "insightvm_scan:1",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "assets": 1,
            "duration": "PT13M40.691S",
            "endTime": "2020-09-24T15:06:38.656Z",
            "engineId": 3,
            "engineName": "Local scan engine",
            "id": 1,
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/scans/1",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/scan_engines/3",
                "id": 3,
                "rel": "Scan Engine",
              },
              Object {
                "href": "https://localhost:3780/api/3/sites/1",
                "rel": "Site",
              },
            ],
            "scanName": "First Scan",
            "scanType": "Manual",
            "siteId": 1,
            "siteName": "Local Site",
            "startTime": "2020-09-24T14:52:57.965Z",
            "status": "finished",
            "vulnerabilities": Object {
              "critical": 0,
              "moderate": 1,
              "severe": 0,
              "total": 1,
            },
          },
        },
      ],
      "_type": "insightvm_scan",
      "active": false,
      "assets": 1,
      "category": "Vulnerability Scan",
      "completedOn": 1600959998656,
      "createdOn": undefined,
      "displayName": "First Scan",
      "engineId": 3,
      "engineName": "Local scan engine",
      "id": "1",
      "internal": true,
      "name": "First Scan",
      "siteId": 1,
      "siteName": "Local Site",
      "startedOn": 1600959177965,
      "state": "finished",
      "status": "finished",
      "summary": "First Scan",
      "webLink": "https://localhost:3780/api/3/scans/1",
    },
    Object {
      "_class": Array [
        "Device",
      ],
      "_key": "insightvm_asset:1",
      "_rawData": Array [],
      "_type": "insightvm_asset",
      "category": "server",
      "createdOn": undefined,
      "displayName": "extreme-desktop",
      "id": "1",
      "make": "unknown",
      "model": "unknown",
      "name": "extreme-desktop",
      "serial": "unknown",
      "webLink": "https://localhost:3780/api/3/assets/1",
    },
    Object {
      "_class": Array [
        "Finding",
      ],
      "_key": "insightvm_asset_vulnerability:1:unix-user-home-dir-mode",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "unix-user-home-dir-mode",
            "instances": 1,
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/assets/1/vulnerabilities/unix-user-home-dir-mode",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/vulnerabilities/unix-user-home-dir-mode",
                "id": "unix-user-home-dir-mode",
                "rel": "Vulnerability",
              },
              Object {
                "href": "https://localhost:3780/api/3/assets/1/vulnerabilities/unix-user-home-dir-mode/validations",
                "id": "unix-user-home-dir-mode",
                "rel": "Vulnerability Validations",
              },
              Object {
                "href": "https://localhost:3780/api/3/assets/1/vulnerabilities/unix-user-home-dir-mode/solution",
                "id": "unix-user-home-dir-mode",
                "rel": "Vulnerability Solutions",
              },
            ],
            "results": Array [
              Object {
                "proof": "<p><p>The permissions for home directory of user extreme was found to be 755 which is more permissive than 750.</p></p>",
                "since": "2020-09-24T15:06:28.045Z",
                "status": "vulnerable-version",
              },
            ],
            "since": "2020-09-24T15:06:28.045Z",
            "status": "vulnerable",
          },
        },
      ],
      "_type": "insightvm_finding",
      "active": undefined,
      "category": "host",
      "createdOn": undefined,
      "displayName": "unix-user-home-dir-mode",
      "id": "unix-user-home-dir-mode",
      "name": "unix-user-home-dir-mode",
      "numericSeverity": 5,
      "open": true,
      "severity": "unknown",
      "status": "vulnerable",
    },
    Object {
      "_class": Array [
        "Vulnerability",
      ],
      "_key": "insightvm_vulnerability:unix-user-home-dir-mode",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "unix-user-home-dir-mode",
            "instances": 1,
            "links": Array [
              Object {
                "href": "https://localhost:3780/api/3/assets/1/vulnerabilities/unix-user-home-dir-mode",
                "rel": "self",
              },
              Object {
                "href": "https://localhost:3780/api/3/vulnerabilities/unix-user-home-dir-mode",
                "id": "unix-user-home-dir-mode",
                "rel": "Vulnerability",
              },
              Object {
                "href": "https://localhost:3780/api/3/assets/1/vulnerabilities/unix-user-home-dir-mode/validations",
                "id": "unix-user-home-dir-mode",
                "rel": "Vulnerability Validations",
              },
              Object {
                "href": "https://localhost:3780/api/3/assets/1/vulnerabilities/unix-user-home-dir-mode/solution",
                "id": "unix-user-home-dir-mode",
                "rel": "Vulnerability Solutions",
              },
            ],
            "results": Array [
              Object {
                "proof": "<p><p>The permissions for home directory of user extreme was found to be 755 which is more permissive than 750.</p></p>",
                "since": "2020-09-24T15:06:28.045Z",
                "status": "vulnerable-version",
              },
            ],
            "since": "2020-09-24T15:06:28.045Z",
            "status": "vulnerable",
          },
        },
      ],
      "_type": "insightvm_vulnerability",
      "active": undefined,
      "blocking": false,
      "category": "other",
      "createdOn": undefined,
      "displayName": "unix-user-home-dir-mode",
      "id": "unix-user-home-dir-mode",
      "name": "unix-user-home-dir-mode",
      "open": false,
      "production": false,
      "public": true,
      "severity": "critical",
      "status": "vulnerable",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "insightvm_account:root",
      "_key": "insightvm_account:root|has|insightvm_user:1",
      "_toEntityKey": "insightvm_user:1",
      "_type": "insightvm_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "insightvm_account:root",
      "_key": "insightvm_account:root|has|insightvm_user:2",
      "_toEntityKey": "insightvm_user:2",
      "_type": "insightvm_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "insightvm_account:root",
      "_key": "insightvm_account:root|has|insightvm_user:3",
      "_toEntityKey": "insightvm_user:3",
      "_type": "insightvm_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "insightvm_account:root",
      "_key": "insightvm_account:root|has|insightvm_site:1",
      "_toEntityKey": "insightvm_site:1",
      "_type": "insightvm_account_has_site",
      "displayName": "HAS",
    },
    Object {
      "_class": "PERFORMED",
      "_fromEntityKey": "insightvm_site:1",
      "_key": "insightvm_site:1|performed|insightvm_scan:1",
      "_toEntityKey": "insightvm_scan:1",
      "_type": "insightvm_site_performed_scan",
      "displayName": "PERFORMED",
    },
    Object {
      "_class": "OWNS",
      "_fromEntityKey": "insightvm_user:2",
      "_key": "insightvm_user:2|owns|insightvm_asset:1",
      "_toEntityKey": "insightvm_asset:1",
      "_type": "insightvm_user_owns_asset",
      "displayName": "OWNS",
    },
    Object {
      "_class": "OWNS",
      "_fromEntityKey": "insightvm_user:1",
      "_key": "insightvm_user:1|owns|insightvm_asset:1",
      "_toEntityKey": "insightvm_asset:1",
      "_type": "insightvm_user_owns_asset",
      "displayName": "OWNS",
    },
    Object {
      "_class": "OWNS",
      "_fromEntityKey": "insightvm_user:3",
      "_key": "insightvm_user:3|owns|insightvm_asset:1",
      "_toEntityKey": "insightvm_asset:1",
      "_type": "insightvm_user_owns_asset",
      "displayName": "OWNS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "insightvm_site:1",
      "_key": "insightvm_site:1|has|insightvm_asset:1",
      "_toEntityKey": "insightvm_asset:1",
      "_type": "insightvm_site_has_asset",
      "displayName": "HAS",
    },
    Object {
      "_class": "MONITORS",
      "_fromEntityKey": "insightvm_scan:1",
      "_key": "insightvm_scan:1|monitors|insightvm_asset:1",
      "_toEntityKey": "insightvm_asset:1",
      "_type": "insightvm_scan_monitors_asset",
      "displayName": "MONITORS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "insightvm_asset:1",
      "_key": "insightvm_asset:1|has|insightvm_asset_vulnerability:1:unix-user-home-dir-mode",
      "_toEntityKey": "insightvm_asset_vulnerability:1:unix-user-home-dir-mode",
      "_type": "insightvm_asset_has_finding",
      "displayName": "HAS",
    },
    Object {
      "_class": "IS",
      "_fromEntityKey": "insightvm_asset_vulnerability:1:unix-user-home-dir-mode",
      "_key": "insightvm_asset_vulnerability:1:unix-user-home-dir-mode|is|insightvm_vulnerability:unix-user-home-dir-mode",
      "_toEntityKey": "insightvm_vulnerability:unix-user-home-dir-mode",
      "_type": "insightvm_finding_is_vulnerability",
      "displayName": "IS",
    },
  ],
  "encounteredTypes": Array [
    "insightvm_account",
    "insightvm_user",
    "insightvm_account_has_user",
    "insightvm_site",
    "insightvm_account_has_site",
    "insightvm_scan",
    "insightvm_site_performed_scan",
    "insightvm_asset",
    "insightvm_user_owns_asset",
    "insightvm_site_has_asset",
    "insightvm_scan_monitors_asset",
    "insightvm_finding",
    "insightvm_asset_has_finding",
    "insightvm_vulnerability",
    "insightvm_finding_is_vulnerability",
  ],
  "numCollectedEntities": 9,
  "numCollectedRelationships": 12,
}
`;
